### PR TITLE
Hack around with reorganizing trends/breakdown code

### PR DESCRIPTION
**Pull request by macobo**
Did a bit of refactoring on trends to see how we could organize the code. The idea is that with these blocks we can do better "unit" testing on smaller units which do less things.

All is in one file for now - this is not intended to be merged but to spark a discussion

- Dont silence errors during test
- Add an explicit array cast to breakdown array value
- Add another hacky shortcut for breaking down by no props
- Remove unneeded indirection
- Move shared code out of if statement
- Remove copy-pasted code
- extract_method into separate class
- Extract method: base_query_arguments
- Extract breakdown filter building to a separate class
- Make linter happy
- Reorganize code
- Make breakdown code into a class
- Extract code into methods or into constructor
- Move code around
- Extract classes, separate the code

## Changes

*Please describe.*  
*If this affects the frontend, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: ee/clickhouse/queries/trends/breakdown.py [hunk 1]
```diff
@@ -234,41 +119,140 @@ def _determine_breakdown_label(
...
 
-    def _get_top_elements(self, query: str, filter: Filter, team_id: int) -> List:
-        element_params = {"key": filter.breakdown, "limit": 20, "team_id": team_id}
 
-        try:
-            top_elements_array_result = sync_execute(query, element_params)
-            top_elements_array = top_elements_array_result[0][0]
-        except:
-            top_elements_array = []
+class BreakdownAggregateTrends(ClickhouseTrendsBreakdown):
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/breakdown.py:123]: This pattern is already alot clearer than the long if/else blocks from before

File: ee/clickhouse/queries/trends/breakdown.py [hunk 2]
```diff
@@ -234,41 +119,140 @@ def _determine_breakdown_label(
...
+    @cached_property
+    def built_query(self):
+        return self.build_query()
+
+    @abstractmethod
+    def build_query(self):
+        pass
+
+
+class CohortBreakdownFilterConstructor(BreakdownFilterConstructor):
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/breakdown.py:240]: I like these filter constructors. This pattern can be applied to other clickhouse queries like lifecycle to easily extend current functionality

**Review by EDsCODE — COMMENTED** [review]
Overall, I think this is much more organized than the previous iteration. I'm curious if we can get some sort of similar format for the postgres side of this query to promote higher consistency. That's generally been the biggest pain point for me—complete structure switching when implementing each version of the query

**Comment by macobo** [comment]
Updated the branch.

> I'm curious if we can get some sort of similar format for the postgres side of this query to promote higher consistency.

I think so, but this requires a bit of extra work. Are you OK if I handle this next time I'm working on trends code which affects both in a deep way? The correct way to achieve similarity is not obvious right now - e.g. breakdowns code is super coupled with the rest of the trends in postgres side of things.

**Pull request closed without merging**